File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/parts/ExponentPart.java
package xyz.wagyourtail.commons.parsers.impl.constant.parts;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.ArrayList;
import java.util.List;

public class ExponentPart extends StringData.OnlyRaw<Data.ListContent> {

    public ExponentPart(String rawContent) {
        super(rawContent, ExponentPart::getContentChecked);
    }

    public ExponentPart(CharReader<?> reader) {
        super(reader, ExponentPart::getContentChecked);
    }

    public static ExponentPart parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new ExponentPart(reader);
        reader.expectEOS();
        return number;
    }

    public static ExponentPart unchecked(String rawContent) {
        return new ExponentPart(rawContent);
    }

    private static ListContent getContentChecked(CharReader<?> reader) {
        List<Object> content = new ArrayList<>(2);

        val first = reader.peek();
        if (first == '-' || first == '+') {
            content.add((char) reader.take());
        }

        content.add(new DecimalPart(reader));

        return new ListContent(content);
    }
}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/number/ExponentPart.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant.number

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class ExponentPart(rawContent: String) : StringData.OnlyRaw<Data.ListContent>(rawContent) {

    constructor(content: Data.ListContent) : this(content.toString())

    companion object : StringDataBuilder<ExponentPart, Data.ListContent> {

        override fun invoke(rawContent: CharReader<*>): ExponentPart {
            return ExponentPart(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): ListContent {
            val content = mutableListOf<Any>()

            val first = reader.peek()
            if (first == '-' || first == '+') {
                content.add(reader.take()!!)
            }

            content.add(DecimalPart(reader))

            return ListContent(content)
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): ListContent {
        return Companion.checkedBuildContent(reader)
    }

}